radix-engine/src/blueprints/account/blueprint.rs

```diff
@@ -350,77 +348,104 @@ impl AccountBlueprint {
         Y: ClientApi<RuntimeError>,
     {
         let resource_address = bucket.resource_address(api)?;
-        let deposits_mode = Self::get_current_deposits_mode(api)?;
 
-        let is_deposit_allowed = Self::is_deposit_allowed(&deposits_mode, &resource_address, api)?;
-        if !is_deposit_allowed {
-            return Ok(Some(bucket));
+        let is_deposit_allowed = Self::is_deposit_allowed(&resource_address, api)?;
+        if is_deposit_allowed {
+            Self::get_vault(
+                resource_address,
+                |vault, api| vault.put(bucket, api),
+                true,
+                api,
+            )?;
+            Ok(None)
+        } else {
+            Ok(Some(bucket))
         }
+    }
 
-        Self::get_vault(
-            resource_address,
-            |vault, api| vault.put(bucket, api),
-            true,
-            api,
-        )?;
+    /// Method is public to all - if ANY of the resources can't be deposited then ALL are returned.
+    pub fn try_deposit_batch_return_on_failure<Y>(
+        buckets: Vec<Bucket>,
+        api: &mut Y,
+    ) -> Result<Vec<Bucket>, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        let can_all_be_deposited = buckets
+            .iter()
+            .map(|bucket| {
+                bucket
+                    .resource_address(api)
+                    .and_then(|resource_address| Self::is_deposit_allowed(&resource_address, api))
+            })
+            .all(|item| item == Ok(true));
 
-        Ok(None)
+        if can_all_be_deposited {
+            Self::deposit_batch(buckets, api)?;
+            Ok(vec![])
+        } else {
+            Ok(buckets)
+        }
     }
 
-    pub fn safe_deposit_batch<Y>(
+    /// Method is public to all - if the resources can't be deposited then the execution panics.
+    pub fn try_deposit_abort_on_failure<Y>(bucket: Bucket, api: &mut Y) -> Result<(), RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        if let Some(bucket) = Self::try_deposit(bucket, api)? {
+            let resource_address = bucket.resource_address(api)?;
+            Err(AccountError::DepositIsDisallowed { resource_address }.into())
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Method is public to all - if ANY of the resources can't be deposited then the execution
+    /// panics.
+    pub fn try_deposit_batch_abort_on_failure<Y>(
         buckets: Vec<Bucket>,
         api: &mut Y,
-    ) -> Result<Vec<Bucket>, RuntimeError>
+    ) -> Result<(), RuntimeError>
     where
         Y: ClientApi<RuntimeError>,
     {
-        let mut undeposited_buckets = vec![];
         for bucket in buckets {
-            let rtn = Self::safe_deposit(bucket, api)?;
-            if let Some(bucket) = rtn {
-                undeposited_buckets.push(bucket)
-            }
+            Self::try_deposit_abort_on_failure(bucket, api)?;
         }
-
-        Ok(undeposited_buckets)
+        Ok(())
```

Comment:
It would be efficient to call `try_deposit_batch_return_on_failure` here, because it check deposit rules upfront?